File: src/main/java/com/esentri/wrabbitmq/java/WrabbitEventWithReply.java
package com.esentri.wrabbitmq.java;

import com.esentri.wrabbitmq.WrabbitTopic;
import kotlin.Unit;
import kotlin.jvm.functions.Function2;

import java.io.Serializable;
import java.util.Map;
import java.util.UUID;
import java.util.function.BiConsumer;
import java.util.function.BiFunction;
import java.util.function.Consumer;
import java.util.function.Function;

public class WrabbitEventWithReply<MESSAGE extends Serializable, REPLY extends Serializable> extends com.esentri.wrabbitmq.WrabbitEventWithReply<MESSAGE, REPLY> {
   public WrabbitEventWithReply(WrabbitTopic wrabbitTopic, String eventName) {
      super(wrabbitTopic, eventName);
   }

   public void listener(Consumer<MESSAGE> listener) {
      this.listener(UUID.randomUUID().toString(), it -> {
         listener.accept(it);
         return Unit.INSTANCE;
      });
   }

   public void listener(String group, Consumer<MESSAGE> listener) {
      this.listener(group, it -> {
         listener.accept(it);
         return Unit.INSTANCE;
      });
   }

   public void listener(BiConsumer<Map<String, ?>, MESSAGE> listener) {
      this.listener(UUID.randomUUID().toString(), listener);
   }

   public void listener(String group, BiConsumer<Map<String, ?>, MESSAGE> listener) {
      this.listener(group, (context, it) -> {
         listener.accept(context, it);
         return Unit.INSTANCE;
      });
   }

   public void replier2(BiFunction<Map<String, ?>, MESSAGE, REPLY> replier) {
      Function2<Map<String, ?>, MESSAGE, REPLY> function2 = replier::apply;
      this.replier(function2);
   }
}

File: src/main/kotlin/com/esentri/wrabbitmq/WrabbitEventWithReply.kt
package com.esentri.wrabbitmq

import com.esentri.wrabbitmq.connection.WrabbitHeader
import com.esentri.wrabbitmq.internal.SendAndReceiveMessage
import com.esentri.wrabbitmq.internal.consumer.WrabbitConsumerReplier
import java.io.Serializable
import java.util.*
import java.util.concurrent.CompletableFuture

open class WrabbitEventWithReply<MESSAGE: Serializable, RETURN: Serializable>(
   wrabbitTopic: WrabbitTopic,
   eventName: String): WrabbitEvent<MESSAGE>(wrabbitTopic, eventName) {

   override fun messageBuilder() = WrabbitMessageBuilderReplier<MESSAGE, RETURN>(wrabbitTopic.topicName, super.standardSendingProperties)

   @JvmOverloads
   fun sendAndReceive(message: MESSAGE, timeoutMS: Long = WrabbitReplyTimeoutMS()): CompletableFuture<RETURN> =
      SendAndReceiveMessage(wrabbitTopic.topicName,
         super.standardSendingProperties,
         message,
         timeoutMS)

   fun replier(replier: WrabbitReplier<MESSAGE, RETURN>) {
      this.replier { _, message ->  replier(message)}
   }

   fun replier(replier: WrabbitReplierWithContext<MESSAGE, RETURN>) {
      val newChannel = ThreadChannel()
      val queueName = "${wrabbitTopic.topicName}.$eventName.REPLIER"
      newChannel.queueDeclare(queueName, true, true, false, emptyMap())
      newChannel.queueBind(queueName, wrabbitTopic.topicName, "", replierHeadersForEvent())
      newChannel.basicConsume(queueName, false, WrabbitConsumerReplier(newChannel, replier))
   }

   private fun replierHeadersForEvent(): MutableMap<String, Any?> {
      val headers: MutableMap<String, Any?> = HashMap()
      headers["x-match"] = "all"
      headers[WrabbitHeader.EVENT.key] = eventName
      return headers
   }

}